From effcd70a980b8c4d6f4d82d04b79972832863267 Mon Sep 17 00:00:00 2001
From: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Date: Fri, 22 Dec 2017 19:37:14 +0300
Subject: [PATCH 1/2] ARC: HSDK: patch linux dtb according to core_mask

Patch cpus 'status' property in linux dtb according to core_mask

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
---
 board/synopsys/hsdk/hsdk-cmd.c | 26 +++++++++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/board/synopsys/hsdk/hsdk-cmd.c b/board/synopsys/hsdk/hsdk-cmd.c
index b706901d6c..c2d5ab0afa 100644
--- a/board/synopsys/hsdk/hsdk-cmd.c
+++ b/board/synopsys/hsdk/hsdk-cmd.c
@@ -8,6 +8,7 @@
 #include <linux/kernel.h>
 #include <linux/io.h>
 #include <asm/arcregs.h>
+#include <fdt_support.h>
 
 
 #ifdef CONFIG_CPU_BIG_ENDIAN
@@ -20,6 +21,7 @@
 
 /* TODO: move to common config */
 #define NR_CPUS		4
+#define ALL_CPU_MASK	0xF	/* GENMASK(NR_CPUS, 0) */
 #define MASTER_CPU	0
 #define MAX_CMD_LEN	25
 #define HZ_IN_MHZ	1000000
@@ -936,7 +938,9 @@ static int hsdk_go_run(u32 cpu_start_reg)
 
 int board_prep_linux(bootm_headers_t *images)
 {
+	u32 i;
 	int ret;
+	char dt_cpu_path[30];
 
 	ret = env_read_validate_common(env_map_mask);
 	if (ret)
@@ -944,12 +948,32 @@ int board_prep_linux(bootm_headers_t *images)
 
 	/* Rollback to default values */
 	if (!env_common.core_mask.set) {
-		env_common.core_mask.val = 0xF;
+		env_common.core_mask.val = ALL_CPU_MASK;
 		env_common.core_mask.set = true;
 	}
 
 	printf("CPU start mask is %#x\n", env_common.core_mask.val);
 
+	if (!is_cpu_used(MASTER_CPU))
+		pr_err("ERR: try to launch linux with CPU[0] disabled! It doesn't work for ARC.\n");
+
+	if (!IMAGE_ENABLE_OF_LIBFDT || !images->ft_len) {
+		if (env_common.core_mask.val != ALL_CPU_MASK) {
+			pr_err("WARN: core_mask setup will work properly only with external DTB!\n");
+
+			return 0;
+		}
+	}
+
+	for (i = 0; i < NR_CPUS; i++) {
+		if (!is_cpu_used(i)) {
+			sprintf(dt_cpu_path, "/cpus/cpu@%u", i);
+			ret = fdt_status_disabled_by_alias(images->ft_addr, dt_cpu_path);
+			debug("patched '%s' node status: ret=%d%s\n", dt_cpu_path,
+			      ret, ret == 0 ? "(OK)" : "");
+		}
+	}
+
 	return 0;
 }
 
-- 
2.11.0

